src/Oscoin/Storage/Block/Cache.hs

```diff
@@ -0,0 +1,66 @@
+{-- | A cache for 'Blocks', implemented as a ring buffer.
+ -- It caches the \"mutable\" part of the chain, which are usually the
+ -- last @k@ blocks, where @k@ is a protocol constant.
+ --}
+module Oscoin.Storage.Block.Cache
+    ( BlockCache
+    , newBlockCache
+    , consBlock
+    , viewBlocks
+    , invalidate
+    , cached
+    ) where
+
+import           Oscoin.Prelude hiding (length)
+
+import           GHC.Natural
+
+import           Oscoin.Crypto.Blockchain.Block (Block)
+import           Oscoin.Time.Chrono (NewestFirst(..))
+
+import qualified Data.Foldable as F
+import           Data.Sequence.Circular (CircularSeq)
+import qualified Data.Sequence.Circular as CSeq
+
+newtype BlockCache c tx s =
+    BlockCache { getBlockCache :: MVar (CircularSeq (Block c tx s)) }
+
+cached
+    :: BlockCache c tx s
+    -> Int
+    -- ^ The number of requested elements
+    -> (Int -> IO (NewestFirst [] (Block c tx s)))
+    -- ^ A function to \"backfill\" the cache by the number of missing
+    -- elements.
+    -> IO (NewestFirst [] (Block c tx s))
+cached (BlockCache bc) requestedElements backFill =
+    modifyMVar bc $ \cseq -> do
+        let currentSize = CSeq.getSize cseq
+        if currentSize < requestedElements
+          then do
+              newElements <- backFill (requestedElements - currentSize)
+              let cseq' = foldr (CSeq.<|) cseq newElements
+              pure (cseq', NF $ take requestedElements $ F.toList cseq')
```

Comment:
The `NF` thing hasn't been changed - I don't know if you intended to or not?

Reply:
Left it out on purpose whilst we decided on the fate of the name of the data structures on `Oscoin.Time.Chrono` to not embark into another "search & replace", but looks like now is time to do that 😉

Reply:
a42f62e